### ISSUE TITLE
Support Swift Package Manager projects

### ISSUE THREAD

**Issue by rfdickerson** [codecov/support#142]
Now that Swift has been open-sourced and cross-platform, we would like CodeCov support on Swift projects that use Swift Package Manager and the open source version of [XCTests](https://github.com/apple/swift-corelibs-xctest). Currently, Codecov requires xcodebuild to set up testcase coverage.

**Comment by stevepeak** [comment]
Hey @rfdickerson I'm not sure how to accommodate this request. 

@bgerstle @coreyfloyd could you help out here? Thank you 👍

**Comment by bgerstle** [comment]
I agree this would be nice, but I don't have a firm understanding of the Swift toolchain.  Seems like @rfdickerson has made some progress at https://github.com/IBM-Swift/Kitura/issues/137, but appears to be stuck on generating the coverage logs.  

@rfdickerson is there something specific to codecov that's blocking you, or just figuring stuff out with the Swift toolchain itself? If so, @stevepeak might be able to help.  Otherwise, is there something else in particular we can assist you with?  

I will follow-up here if I'm able to dedicate any time to investigating this further.  Like I said, I haven't dug into the Swift toolchain yet, so keep in mind I'd have to overcome the learning curve before making a dent in the problem at hand.

**Comment by bgerstle** [comment]
@rfdickerson are there any other projects getting coverage data using the Swift toolchain?  I'd imagine that Swift team would probably like to have coverage reports for the language itself, not to mention the standard library!

**Comment by paulofaria** [comment]
Hey! just dropping by to let you guys know that I'm also very interested it this integration.

**Comment by rfdickerson** [comment]
One major stumbling block is that we can't get the normal Code coverage working correctly in XCode using the open source Swift toolchain. The reason is that in XCode the LLVM tools like `llvm-profdata` and `llvm-cov` are the old version, and not the version that the swift toolchain is using, a [branched version](https://github.com/apple/swift-llvm) of 3.9. 

In anycase, I have been able to use the command-line to produce coverage and profiling information using:

```
swift build -Xswiftc -profile-generate -Xswiftc -profile-coverage-mapping
```

Then transform the .profraw file to a profdata file using the Apple's fork of LLVM:

```
/Ninja-ReleaseAssert/llvm-macosx-x86_64/bin/llvm-profdata merge -o testprof.profdata default.profraw
```

It looks like your scripts require an .xccoverage file in the Derived Data directory? Any advise in producing that xccoverage file?

**Comment by bgerstle** [comment]
@stevepeak correct me if I'm wrong, but I believe it uses `llvm-show` to produce a report, and parses the data from there. I would also try [slather](https://github.com/SlatherOrg/slather) to see if they're able to parse the coverage data & upload to codecov.

**Comment by rfdickerson** [comment]
I have been following the section in [this tutorial from Nerd Ranch](https://www.bignerdranch.com/blog/weve-got-you-covered/):

```
$ xcrun swiftc -profile-generate -profile-coverage-mapping testprof.swift
$ xcrun llvm-profdata merge -o testprof.profdata default.profraw
$ xcrun llvm-cov show ./testprof -instr-profile=testprof.profdata testprof.m
```

In my case, would the testprof.m be a particular Swift file that I want code codecoverage report for? The following still gives me an error.

**Comment by bgerstle** [comment]
Maybe you can upload some of the coverage data artifacts you're trying to
process to the issue so @stevepeak can hack on them?

**Comment by stevepeak** [comment]
@rfdickerson I'm happy to check out the file! Thank you @bgerstle 

Pass me a owner/repo@sha and I'll take a look. Thank you 👍

**Comment by rfdickerson** [comment]
Hi @stevepeak , that would be amazing if you could look it over at your availability. The repository is a Swift 3. 04-12, 4-25, 05-03 compatible project:

https://github.com/rfdickerson/SwiftDeflate

**Comment by ilazakis** [comment]
Hey @stevepeak! Any updates on this issue? It would be really useful to have this as more and more "pure" Swift packages surface i.e. packages that are not accompanied by an Xcode project. Thanks!

**Comment by stevepeak** [comment]
@rfdickerson Thanks for the example repo. When I get a chance we can test against it.
@ilazakis no update at the moment 😦 we are kindly pointing the finger at Swift atm and will inquire with the Swift team concerning the unexpected 0% coverage output.

**Comment by stevepeak** [comment]
> deleted a post concerning another issue

I'll look into this now to see where we left off and open a Beta for everyone to test against. Thank you for your patience and help 👍

**Comment by ilazakis** [comment]
Thanks for the swift reply 👍 @stevepeak 😃

**Comment by stevepeak** [comment]
> ```sh
# (hopefully) this is the magic sauce
xcodebuild -showBuildSettings 2>/dev/null | grep -i "^\s*PRODUCT_NAME" | sed -e 's/.*= \(.*\)/\1/'
```

To **beta** please update your bash uploader temporarily to:

```sh
bash <(curl -s https://raw.githubusercontent.com/codecov/codecov-bash/feature/swift/package_detection/codecov)
```

Please post your CI build results below 👍 🎉 💯

**Comment by ilazakis** [comment]
Thanks for the input @stevepeak !

Looks like the main issue with "pure" Swift packages is that tests run using `swift test`  instead of `xcodebuild` to compile. It looks like ````swift test```` does not produce any coverage data (could be wrong).

That said, I was able to get a coverage report by taking advantage of Swift Package Manager's `swift package generate-xcodeproj` command which generates an Xcode project for that package.

That command, followed by an `xcodebuild` test build works just fine. We can then follow up with a cleanup of the project and build artifacts.

An example snippet of a `travis` setup that worked for me:

````
language: generic
sudo: required
osx_image: xcode8
script:
  - swift package generate-xcodeproj
  - xcodebuild -scheme RequestBuilder -enableCodeCoverage YES test  # instead of swift test
after_success:
  - bash <(curl -s https://codecov.io/bash)
````

[And the successful result](https://codecov.io/gh/ilazakis/BuildIt/commit/6f491a3859b95d17aab3ba4ca56b732983cf67d0)

**Comment by stevepeak** [comment]
@ilazakis correct me if I'm wrong but I think your point above is more related to https://github.com/codecov/example-swift/issues/14

This issue is concerning auto-detect the `-J AppName` setting by inspecting the build settings. So that CocoPods and other dependencies are not processed (which can be a time consuming activity).

**Comment by NicholasTD07** [comment]
Beta script failed to find codecov report with `swift test`.

**Travis CI build log**: https://travis-ci.org/NicholasTD07/ExpenseKit/builds/163081681


**`.travis.yml`**: https://github.com/NicholasTD07/ExpenseKit/blob/37683d52daaec85e6d2b8987dd90f9f8dc0a967e/.travis.yml

```
17.91s $ swift test
Compile Swift Module 'ExpenseKit' (1 sources)
Compile Swift Module 'ExpenseKitTests' (1 sources)
Linking ./.build/debug/ExpenseKitPackageTests.xctest/Contents/MacOS/ExpenseKitPackageTests
Test Suite 'All tests' started at 2016-09-27 12:23:39.065
Test Suite 'ExpenseKitPackageTests.xctest' started at 2016-09-27 12:23:39.066
Test Suite 'ExpenseKitTests' started at 2016-09-27 12:23:39.067
Test Case '-[ExpenseKitTests.ExpenseKitTests testAddingExpenses]' started.
Test Case '-[ExpenseKitTests.ExpenseKitTests testAddingExpenses]' passed (0.003 seconds).
Test Case '-[ExpenseKitTests.ExpenseKitTests testCreatingBudget]' started.
Test Case '-[ExpenseKitTests.ExpenseKitTests testCreatingBudget]' passed (0.000 seconds).
Test Suite 'ExpenseKitTests' passed at 2016-09-27 12:23:39.072.
	 Executed 2 tests, with 0 failures (0 unexpected) in 0.004 (0.005) seconds
Test Suite 'ExpenseKitPackageTests.xctest' passed at 2016-09-27 12:23:39.072.
	 Executed 2 tests, with 0 failures (0 unexpected) in 0.004 (0.006) seconds
Test Suite 'All tests' passed at 2016-09-27 12:23:39.073.
	 Executed 2 tests, with 0 failures (0 unexpected) in 0.004 (0.007) seconds
The command "swift test" exited with 0.
after_success
0.78s $ bash <(curl -s https://raw.githubusercontent.com/codecov/codecov-bash/feature/swift/package_detection/codecov)
  _____          _
 / ____|        | |
| |     ___   __| | ___  ___ _____   __
| |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
| |___| (_) | (_| |  __/ (_| (_) \ V /
 \_____\___/ \__,_|\___|\___\___/ \_/
                              Bash-tbd
==> Travis CI detected.
    project root: .
    Yaml not found, that's ok! Learn more at http://docs.codecov.io/docs/codecov-yaml
==> Running gcov in . (disable via -X gcov)
==> Searching for coverage reports in:
    + .
==> Python coveragepy not found
--> No coverage report found.
    Please visit http://docs.codecov.io/docs/supported-languages
```

**Comment by stevepeak** [comment]
@NicholasTD07 You'll still need to run the coverage as you had previously here: https://github.com/NicholasTD07/ExpenseKit/blob/e4c3d2457912c319320c76576ed672508242886c/.travis.yml#L8-L9

The only change should be the bash uploader to beta this feature. Thank you!

**Comment by stevepeak** [comment]
@NicholasTD07 You'll still need to run the coverage as you had previously here: https://github.com/NicholasTD07/ExpenseKit/blob/e4c3d2457912c319320c76576ed672508242886c/.travis.yml#L8-L9

The only change should be the bash uploader to beta this feature. Thank you!
